### PR TITLE
Docs updates -- minor changes

### PR DESCRIPTION
**Pull request by sydney-runkle**
Fix https://github.com/pydantic/pydantic/issues/8620
Fix https://github.com/pydantic/pydantic/issues/8591
Fix https://github.com/pydantic/pydantic/issues/8529

### CODE REVIEW

**Comment by Viicos** [comment]
@sydney-runkle, is the return type of `model_dump` always `dict[str, Any]` if the root type is a `BaseModel`? In that case, you could do:

```python
@overload
def model_dump(self: RootModel[BaseModel], ...) -> dict[str, Any]: ...
@overload
def model_dump(self, ...) -> RootModelRootType: ...
```

**Comment by sydney-runkle** [comment]
@Viicos,

Good thought. I think you're right - the concern is that it might not play well with subclasses. Even though it's not a wonderful solution, I think the catch-all `Any` is our best bet for now.

**Comment by Viicos** [comment]
Ok this should work:

```python
RootModelType = TypeVar("RootModelType")

BaseModelT = TypeVar("BaseModelT", bound=BaseModel)


class RootModel(Generic[RootModelType]):
    ...
    @overload
    def model_dump(self: RootModel[BaseModelT], ...) -> dict[str, Any]: ...
    @overload
    def model_dump(self, ...) -> RootModelRootType: ...


class Model(BaseModel):
    pass

r = RootModel(Model)

reveal_type(r.model_dump())  # dict[str, Any]
```

File: pydantic/root_model.py [hunk 1]
```diff
@@ -132,11 +132,17 @@ def model_dump(  # type: ignore
...
+        ) -> Any:
             """This method is included just to get a more accurate return type for type checkers.
             It is included in this `if TYPE_CHECKING:` block since no override is actually necessary.
 
             See the documentation of `BaseModel.model_dump` for more details about the arguments.
+
+            Generally, this method will have a return type of `RootModelRootType`, assuming that `RootModelRootType` is
+            not a `BaseModel` subclass. If `RootModelRootType` is a `BaseModel` subclass, then the return
+            type will likely be `dict[str, Any]`, as `model_dump` calls are recursive. The return type could
+            even be something different, in the case of a custom serializer on a `BaseModel` subclass.
```
Comment (dmontagu) [inline pydantic/root_model.py:144]: or a custom serializer on the root annotation. I think you can just say "in the case of a custom serializer".

Comment (dmontagu) [reply pydantic/root_model.py:144]:
```suggestion
            even be something different, in the case of a custom serializer.
```

**Review by dmontagu — APPROVED** [review]

**Comment by sydney-runkle** [comment]
@Viicos,

I like the thought, but that doesn't account for dumping of `dataclass` types, either. Given that there are so many various cases, we're going to stick with `Any` for the annotation.

That being said, I do think your suggestion is helpful, in the case where individual users of the lib might want to get a more specific return type annotation, especially if their usage of `RootModel` is relatively narrow.

**Comment by Viicos** [comment]
Yeah you're right I thought this would only apply for `BaseModel`'s but as dmontagu said it also applies to anything with a custom serializer.

+ the solution I provided only works under `pyright`, not `mypy`. Better to stick with `Any` (no annotation is better than an erroneous annotation)

**Pull request merged**